src/charm.py

```diff
@@ -278,64 +281,51 @@ def _on_openfga_store_created(self, event: OpenFGAStoreCreateEvent) -> None:
+        self._holistic_handler(event)
 
     def _on_openfga_store_removed(self, event: OpenFGAStoreRemovedEvent) -> None:
         if self.unit.is_leader():
             self.peer_data.pop(key=self._workload_service.version)
 
-        self._handle_status_update_config(event)
+        self._holistic_handler(event)
 
-    @wait_when(container_not_connected)
     def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        if not container_connectivity(self):
+            defer_event(event)
+
         self._workload_service.push_ca_certs(event.ca)
-        self._handle_status_update_config(event)
+        self._holistic_handler(event)
 
-    @wait_when(container_not_connected)
     def _on_certificate_removed(self, event: CertificateRemovedEvent) -> None:
-        self._workload_service.remove_ca_certs()
-        self._handle_status_update_config(event)
-
-    @wait_when(
-        container_not_connected,
-        integration_not_exists(PEER_INTEGRATION_NAME),
-    )
-    @block_when(
-        integration_not_exists(KRATOS_INFO_INTEGRATION_NAME),
-        integration_not_exists(HYDRA_ENDPOINTS_INTEGRATION_NAME),
-        integration_not_exists(OPENFGA_INTEGRATION_NAME),
-        integration_not_exists(INGRESS_INTEGRATION_NAME),
-    )
-    def _handle_status_update_config(self, event: EventBase) -> None:
-        if self.oauth_integration.is_ready() and (
-            not self.model.relations[CERTIFICATE_TRANSFER_INTEGRATION_NAME]
-        ):
-            self.unit.status = BlockedStatus(
-                "Missing certificate_transfer integration with oauth provider"
-            )
-            return
-
-        self.unit.status = MaintenanceStatus("Configuring the Admin Service container")
+        if not container_connectivity(self):
+            defer_event(event)
 
-        self._workload_service.prepare_dir(path=LOG_DIR)
-
-        if not self.openfga_integration.is_store_ready():
-            event.defer()
-            self.unit.status = WaitingStatus("Waiting for OpenFGA store")
-            return
+        self._workload_service.remove_ca_certs()
+        self._holistic_handler(event)
 
-        if not self.peer_data[self._workload_service.version]:
-            event.defer()
-            self.unit.status = WaitingStatus("Waiting for OpenFGA model")
-            return
+    def _on_collect_status(self, event: CollectStatusEvent) -> None:
+        """The central management of the charm operator's status."""
+        for condition, status in CONDITION_STATUS_REGISTRY:
+            if not condition(self):
+                event.add_status(status)
+                return
 
         try:
             self._pebble_service.plan(self._pebble_layer)
         except PebbleError:
-            self.unit.status = BlockedStatus("Failed to plan pebble layer, please check the logs")
+            event.add_status(BlockedStatus("Failed to plan pebble layer, please check the logs"))
             return
```

Comment:
It feels a little weird planning the service in the `_on_collect_status` function, any reason we don't do this in the `_holistic_handler`?

Reply:
Yeah, that was my question to the juju team. They suggested moving status setting to the `CollectStatusEvent` hook handling otherwise the status might be overridden by the logic in the `CollectStatusEvent` hook.